transmission_interface/include/transmission_interface/differential_transmission.hpp

```diff
@@ -0,0 +1,403 @@
+   */
+  std::string get_handles_info() const;
+
+protected:
+  std::vector<double> actuator_reduction_;
+  std::vector<double> joint_reduction_;
+  std::vector<double> joint_offset_;
+
+  std::vector<JointHandle> joint_position_;
+  std::vector<JointHandle> joint_velocity_;
+  std::vector<JointHandle> joint_effort_;
+
+  std::vector<ActuatorHandle> actuator_position_;
+  std::vector<ActuatorHandle> actuator_velocity_;
+  std::vector<ActuatorHandle> actuator_effort_;
+};
+
+inline DifferentialTransmission::DifferentialTransmission(
+  const std::vector<double> & actuator_reduction,
+  const std::vector<double> & joint_reduction,
+  const std::vector<double> & joint_offset)
+: actuator_reduction_(actuator_reduction),
+  joint_reduction_(joint_reduction),
+  joint_offset_(joint_offset)
+{
+  if (num_actuators() != actuator_reduction_.size() ||
+    num_joints() != joint_reduction_.size() ||
+    num_joints() != joint_offset_.size())
+  {
+    throw Exception("Reduction and offset vectors must have size 2.");
+  }
+
+  if (0.0 == actuator_reduction_[0] ||
+    0.0 == actuator_reduction_[1] ||
+    0.0 == joint_reduction_[0] ||
+    0.0 == joint_reduction_[1])
+  {
+    throw Exception("Transmission reduction ratios cannot be zero.");
+  }
+}
+
+
+template<typename T>
+std::string to_string(const std::vector<T> & list)
+{
+  std::stringstream ss;
+  ss << "[";
+  for (const auto & elem : list) {
+    ss << elem << ", ";
+  }
+
+  if (!list.empty()) {
+    ss.seekp(-2, std::ios_base::end);  // remove last ", "
+  }
+  ss << "]";
+  return ss.str();
+}
+
+template<class T>
+std::vector<std::string> get_names(const std::vector<T> & handles)
+{
+  std::set<std::string> names;
+  std::transform(
+    handles.cbegin(), handles.cend(), std::inserter(names, names.end()),
+    [](const auto & handle) {return handle.get_name();});
+  return std::vector<std::string>(names.begin(), names.end());
+}
+
+template<typename T>
+std::vector<T> get_ordered_handles(
+  const std::vector<T> & unordered_handles, const std::vector<std::string> & names,
+  const std::string & interface_type)
+{
+  std::vector<T> result;
+  for (const auto & name : names) {
+    for (auto & handle : unordered_handles) {
+      if ((handle.get_name() == name) &&
+        (handle.get_interface_name() == interface_type) && handle)
+      {
+        result.push_back(handle);
+      }
+    }
+  }
+  return result;
+}
+
+void DifferentialTransmission::configure(
+  const std::vector<JointHandle> & joint_handles,
+  const std::vector<ActuatorHandle> & actuator_handles)
+{
+  if (joint_handles.empty()) {
+    throw Exception("No joint handles were passed in");
+  }
+
+  if (actuator_handles.empty()) {
+    throw Exception("No actuator handles were passed in");
+  }
+
+  const auto joint_names = get_names(joint_handles);
+  if (joint_names.size() != 2) {
+    throw Exception(
+            "There should be exactly two unique joint names but was given " +
+            to_string(joint_names));
```

Comment:
Would this be more consistent?
```suggestion
      "There should be exactly two unique joint names but was given " +
      to_string(joint_names));
```

Reply:
I gave up on trying for format things since we use uncrustify...